lib/kafka/connection.rb

```diff
@@ -88,10 +93,10 @@ def send_request(request)
 
         @correlation_id += 1
 
-        write_request(request, notification)
+        @operation.write_request(request, notification, @correlation_id)
 
         response_class = request.response_class
-        wait_for_response(response_class, notification) unless response_class.nil?
+        @operation.wait_for_response(response_class, notification, @correlation_id) unless response_class.nil?
```

Comment:
I would rather we add low-level `write` and `read` methods to the existing Connection and handle the SASL stuff in Client – ideally using a separate SaslOperation that does the full thing.